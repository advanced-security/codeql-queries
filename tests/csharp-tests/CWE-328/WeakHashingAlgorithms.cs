using System;
using System.IO;
using System.Text;
using System.Runtime.Serialization;
using System.Security.Cryptography;
using System.Security.Permissions;

public class WeakHash
{
    public static string GetMD5Hash(string str)
    {
        MD5CryptoServiceProvider md5 = new MD5CryptoServiceProvider();
        byte[] bytes = ASCIIEncoding.Default.GetBytes(str);
        byte[] encoded = md5.ComputeHash(bytes);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < encoded.Length; i++)
            sb.Append(encoded[i].ToString("x2"));

        return sb.ToString();
    }

    public static string GetMD5Hash2(string str)
    {
        MD5CryptoServiceProvider md5 = (MD5CryptoServiceProvider)System.Security.Cryptography.HashAlgorithm.Create("MD5");
        byte[] bytes = ASCIIEncoding.Default.GetBytes(str);
        byte[] encoded = md5.ComputeHash(bytes);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < encoded.Length; i++)
            sb.Append(encoded[i].ToString("x2"));

        return sb.ToString();
    }

    public static string GetMD5Hash3(string str)
    {
        System.Security.Cryptography.MD5 md5 = System.Security.Cryptography.MD5.Create();
        byte[] bytes = Encoding.Default.GetBytes(str);
        byte[] encoded = md5.ComputeHash(bytes);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < encoded.Length; i++)
            sb.Append(encoded[i].ToString("x2"));

        return sb.ToString();
    }

    public static string GetMD5Hash4(string password)
    {
        System.Security.Cryptography.MD5 md5 = System.Security.Cryptography.MD5.Create();
        byte[] bytes = Encoding.Default.GetBytes(password);
        byte[] encoded = md5.ComputeHash(bytes);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < encoded.Length; i++)
            sb.Append(encoded[i].ToString("x2"));

        return sb.ToString();
    }

    public static string GetSHAHash1(string password)
    {
        System.Security.Cryptography.SHA1 sha1 = System.Security.Cryptography.SHA1.Create();
        byte[] bytes = Encoding.Default.GetBytes(password);
        byte[] encoded = sha1.ComputeHash(bytes);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < encoded.Length; i++)
            sb.Append(encoded[i].ToString("x2"));

        return sb.ToString();
    }

    public static string GetSHAHash2(string password)
    {
        System.Security.Cryptography.SHA512 sha512 = System.Security.Cryptography.SHA512.Create();
        byte[] bytes = Encoding.Default.GetBytes(password);
        byte[] encoded = sha512.ComputeHash(bytes);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < encoded.Length; i++)
            sb.Append(encoded[i].ToString("x2"));

        return sb.ToString();
    }
}
