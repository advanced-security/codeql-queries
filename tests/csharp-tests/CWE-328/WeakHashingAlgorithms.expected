| WeakHashingAlgorithms.cs:12:40:12:69 | object creation of type MD5CryptoServiceProvider | This function creates a new MD5CryptoServiceProvider() object, which uses a cryptographically weak hash algorithm |
| WeakHashingAlgorithms.cs:25:66:25:121 | call to method Create | This function calls the MD5.Create() or HashAlgorithm.Create() method, which uses a cryptographically weak hash algorithm |
| WeakHashingAlgorithms.cs:38:48:38:88 | call to method Create | This function calls the MD5.Create() or HashAlgorithm.Create() method, which uses a cryptographically weak hash algorithm |
| WeakHashingAlgorithms.cs:51:48:51:88 | call to method Create | This function calls the MD5.Create() or HashAlgorithm.Create() method, which uses a cryptographically weak hash algorithm |
